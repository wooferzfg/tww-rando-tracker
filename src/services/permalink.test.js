import Permalink from './permalink';

describe('Permalink', () => {
  describe('OPTIONS', () => {
    test('returns the correct options', () => {
      expect(Permalink.OPTIONS).toMatchSnapshot();
    });
  });

  describe('RANDOMIZE_ENTRANCES_OPTIONS', () => {
    test('returns the correct options', () => {
      expect(Permalink.RANDOMIZE_ENTRANCES_OPTIONS).toMatchSnapshot();
    });
  });

  describe('SWORD_MODE_OPTIONS', () => {
    test('returns the correct options', () => {
      expect(Permalink.SWORD_MODE_OPTIONS).toMatchSnapshot();
    });
  });

  describe('DROPDOWN_OPTIONS', () => {
    test('returns the correct dropdown options', () => {
      expect(Permalink.DROPDOWN_OPTIONS).toMatchSnapshot();
    });
  });

  describe('DEFAULT_PERMALINK', () => {
    test('returns the default options', () => {
      const options = Permalink.decode(Permalink.DEFAULT_PERMALINK);

      expect(options).toMatchSnapshot();
    });
  });

  describe('decode', () => {
    test('decodes a permalink', () => {
      const options = Permalink.decode('MS4xMC4wAHllZXQABwEDAQ2gvg8Q8AAwAAAAAABAAEgG');

      expect(options).toMatchSnapshot();
    });

    test('throws errors for invalid permalinks', () => {
      expect(() => Permalink.decode('')).toThrow();
      expect(() => Permalink.decode('H')).toThrow();
      expect(() => Permalink.decode('AAAA')).toThrow();
      expect(() => Permalink.decode('BBBBBBBBBBBBBBBBBBBBBBB')).toThrow();
      expect(() => Permalink.decode('MS4xMC4wAHllZXQfwijoi1+gIAHEAAAABAAgiAZAA==')).toThrow();
      expect(() => Permalink.decode('VIOEWJAFOEIWAJVEOWAIVJN')).toThrow();
      expect(() => Permalink.decode('vdsccccccccccccccccccccccccccccccccc')).toThrow();
      expect(() => Permalink.decode('AAAAAAAHHHHHHHHHHHHHHHHHHHHHHHHHHHHH')).toThrow();
      expect(() => Permalink.decode('abcdefghijklmnopqrstuvwxyzABCDEFGHIJ')).toThrow();
      expect(() => Permalink.decode('zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz')).toThrow();
    });
  });

  describe('encode', () => {
    let options;
    let permalink;

    beforeEach(() => {
      permalink = 'MS4xMC4wAHllZXQABwEDAQ2gvg8Q8AAwAAAAAABAAEoG';
      options = Permalink.decode(permalink);
    });

    test('encodes a permalink', () => {
      const encodedPermalink = Permalink.encode(options);

      expect(encodedPermalink).toEqual(permalink);
    });
  });
});
