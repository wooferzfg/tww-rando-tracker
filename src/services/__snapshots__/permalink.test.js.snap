// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permalink DEFAULT_PERMALINK returns the default options 1`] = `
{
  "add_shortcut_warps_between_dungeons": false,
  "chest_type_matches_contents": false,
  "cryptic_hints": true,
  "do_not_generate_spoiler_log": false,
  "fishmen_hints": true,
  "hoho_hints": true,
  "instant_text_boxes": true,
  "keylunacy": false,
  "korl_hints": false,
  "num_barren_hints": 0,
  "num_item_hints": 15,
  "num_location_hints": 5,
  "num_path_hints": 0,
  "num_race_mode_dungeons": 4,
  "num_starting_triforce_shards": 0,
  "prioritize_remote_hints": false,
  "progression_battlesquid": false,
  "progression_big_octos_gunboats": false,
  "progression_combat_secret_caves": false,
  "progression_dungeons": true,
  "progression_expensive_purchases": true,
  "progression_eye_reef_chests": false,
  "progression_free_gifts": true,
  "progression_great_fairies": true,
  "progression_island_puzzles": false,
  "progression_long_sidequests": false,
  "progression_mail": false,
  "progression_minigames": false,
  "progression_misc": true,
  "progression_platforms_rafts": false,
  "progression_puzzle_secret_caves": true,
  "progression_savage_labyrinth": false,
  "progression_short_sidequests": false,
  "progression_spoils_trading": false,
  "progression_submarines": false,
  "progression_tingle_chests": false,
  "progression_treasure_charts": false,
  "progression_triforce_charts": false,
  "race_mode": false,
  "randomize_charts": false,
  "randomize_enemies": false,
  "randomize_enemy_palettes": false,
  "randomize_entrances": "Disabled",
  "randomize_music": false,
  "randomize_starting_island": false,
  "remove_music": false,
  "reveal_full_sea_chart": true,
  "seed_name": "A",
  "skip_rematch_bosses": true,
  "starting_gear": {
    "Bait Bag": 0,
    "Ballad of Gales": 1,
    "Beedle's Chart": 0,
    "Bombs": 0,
    "Boomerang": 0,
    "Cabana Deed": 0,
    "Command Melody": 0,
    "DRC Compass": 0,
    "DRC Dungeon Map": 0,
    "Deku Leaf": 0,
    "Delivery Bag": 0,
    "Din's Pearl": 0,
    "ET Compass": 0,
    "ET Dungeon Map": 0,
    "Earth God's Lyric": 0,
    "Empty Bottle": 0,
    "FF Compass": 0,
    "FF Dungeon Map": 0,
    "FW Compass": 0,
    "FW Dungeon Map": 0,
    "Farore's Pearl": 0,
    "Fill-Up Coupon": 0,
    "Ghost Ship Chart": 0,
    "Grappling Hook": 0,
    "Great Fairy Chart": 0,
    "Hero's Charm": 0,
    "Hookshot": 0,
    "Hurricane Spin": 0,
    "Iron Boots": 0,
    "Light Ring Chart": 0,
    "Maggie's Letter": 0,
    "Magic Armor": 0,
    "Moblin's Letter": 0,
    "Nayru's Pearl": 0,
    "Note to Mom": 0,
    "Octo Chart": 0,
    "Platform Chart": 0,
    "Power Bracelets": 0,
    "Progressive Bomb Bag": 0,
    "Progressive Bow": 0,
    "Progressive Magic Meter": 1,
    "Progressive Picto Box": 0,
    "Progressive Quiver": 0,
    "Progressive Shield": 1,
    "Progressive Sword": 0,
    "Progressive Wallet": 0,
    "Secret Cave Chart": 0,
    "Skull Hammer": 0,
    "Song of Passing": 1,
    "Spoils Bag": 0,
    "Submarine Chart": 0,
    "Telescope": 0,
    "Tingle Tuner": 0,
    "Tingle's Chart": 0,
    "TotG Compass": 0,
    "TotG Dungeon Map": 0,
    "WT Compass": 0,
    "WT Dungeon Map": 0,
    "Wind God's Aria": 0,
  },
  "starting_hcs": 0,
  "starting_pohs": 0,
  "swift_sail": true,
  "sword_mode": "Start with Hero's Sword",
  "version": "1.10.0",
}
`;

exports[`Permalink DROPDOWN_OPTIONS returns the correct dropdown options 1`] = `
{
  "num_race_mode_dungeons": [
    1,
    2,
    3,
    4,
    5,
    6,
  ],
  "num_starting_triforce_shards": [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
  ],
  "randomize_entrances": [
    "Disabled",
    "Dungeons",
    "Secret Caves",
    "Dungeons and Secret Caves (Separately)",
    "Dungeons and Secret Caves (Together)",
  ],
  "sword_mode": [
    "Start with Hero's Sword",
    "No Starting Sword",
    "Swordless",
  ],
}
`;

exports[`Permalink OPTIONS returns the correct options 1`] = `
{
  "ADD_SHORTCUT_WARPS_BETWEEN_DUNGEONS": "add_shortcut_warps_between_dungeons",
  "CHEST_TYPE_MATCHES_CONTENTS": "chest_type_matches_contents",
  "CRYPTIC_HINTS": "cryptic_hints",
  "DO_NOT_GENERATE_SPOILER_LOG": "do_not_generate_spoiler_log",
  "FISHMEN_HINTS": "fishmen_hints",
  "HOHO_HINTS": "hoho_hints",
  "INSTANT_TEXT_BOXES": "instant_text_boxes",
  "INVERT_CAMERA_X_AXIS": "invert_camera_x_axis",
  "INVERT_SEA_COMPASS_X_AXIS": "invert_sea_compass_x_axis",
  "KEYLUNACY": "keylunacy",
  "KORL_HINTS": "korl_hints",
  "NUM_BARREN_HINTS": "num_barren_hints",
  "NUM_ITEM_HINTS": "num_item_hints",
  "NUM_LOCATION_HINTS": "num_location_hints",
  "NUM_PATH_HINTS": "num_path_hints",
  "NUM_RACE_MODE_DUNGEONS": "num_race_mode_dungeons",
  "NUM_STARTING_TRIFORCE_SHARDS": "num_starting_triforce_shards",
  "PRIORITIZE_REMOTE_HINTS": "prioritize_remote_hints",
  "PROGRESSION_BATTLESQUID": "progression_battlesquid",
  "PROGRESSION_BIG_OCTOS_GUNBOATS": "progression_big_octos_gunboats",
  "PROGRESSION_COMBAT_SECRET_CAVES": "progression_combat_secret_caves",
  "PROGRESSION_DUNGEONS": "progression_dungeons",
  "PROGRESSION_EXPENSIVE_PURCHASES": "progression_expensive_purchases",
  "PROGRESSION_EYE_REEF_CHESTS": "progression_eye_reef_chests",
  "PROGRESSION_FREE_GIFTS": "progression_free_gifts",
  "PROGRESSION_GREAT_FAIRIES": "progression_great_fairies",
  "PROGRESSION_ISLAND_PUZZLES": "progression_island_puzzles",
  "PROGRESSION_LONG_SIDEQUESTS": "progression_long_sidequests",
  "PROGRESSION_MAIL": "progression_mail",
  "PROGRESSION_MINIGAMES": "progression_minigames",
  "PROGRESSION_MISC": "progression_misc",
  "PROGRESSION_PLATFORMS_RAFTS": "progression_platforms_rafts",
  "PROGRESSION_PUZZLE_SECRET_CAVES": "progression_puzzle_secret_caves",
  "PROGRESSION_SAVAGE_LABYRINTH": "progression_savage_labyrinth",
  "PROGRESSION_SHORT_SIDEQUESTS": "progression_short_sidequests",
  "PROGRESSION_SPOILS_TRADING": "progression_spoils_trading",
  "PROGRESSION_SUBMARINES": "progression_submarines",
  "PROGRESSION_TINGLE_CHESTS": "progression_tingle_chests",
  "PROGRESSION_TREASURE_CHARTS": "progression_treasure_charts",
  "PROGRESSION_TRIFORCE_CHARTS": "progression_triforce_charts",
  "RACE_MODE": "race_mode",
  "RANDOMIZE_CHARTS": "randomize_charts",
  "RANDOMIZE_ENEMIES": "randomize_enemies",
  "RANDOMIZE_ENEMY_PALETTES": "randomize_enemy_palettes",
  "RANDOMIZE_ENTRANCES": "randomize_entrances",
  "RANDOMIZE_MUSIC": "randomize_music",
  "RANDOMIZE_STARTING_ISLAND": "randomize_starting_island",
  "REMOVE_MUSIC": "remove_music",
  "REVEAL_FULL_SEA_CHART": "reveal_full_sea_chart",
  "SEED_NAME": "seed_name",
  "SKIP_REMATCH_BOSSES": "skip_rematch_bosses",
  "STARTING_GEAR": "starting_gear",
  "STARTING_HCS": "starting_hcs",
  "STARTING_POHS": "starting_pohs",
  "SWIFT_SAIL": "swift_sail",
  "SWORD_MODE": "sword_mode",
  "VERSION": "version",
}
`;

exports[`Permalink RANDOMIZE_ENTRANCES_OPTIONS returns the correct options 1`] = `
{
  "DISABLED": "Disabled",
  "DUNGEONS": "Dungeons",
  "DUNGEONS_AND_SECRET_CAVES_SEPARATELY": "Dungeons and Secret Caves (Separately)",
  "DUNGEONS_AND_SECRET_CAVES_TOGETHER": "Dungeons and Secret Caves (Together)",
  "SECRET_CAVES": "Secret Caves",
}
`;

exports[`Permalink SWORD_MODE_OPTIONS returns the correct options 1`] = `
{
  "NO_STARTING_SWORD": "No Starting Sword",
  "START_WITH_HEROS_SWORD": "Start with Hero's Sword",
  "SWORDLESS": "Swordless",
}
`;

exports[`Permalink decode decodes a permalink 1`] = `
{
  "add_shortcut_warps_between_dungeons": false,
  "chest_type_matches_contents": false,
  "cryptic_hints": true,
  "do_not_generate_spoiler_log": false,
  "fishmen_hints": true,
  "hoho_hints": true,
  "instant_text_boxes": true,
  "keylunacy": false,
  "korl_hints": false,
  "num_barren_hints": 0,
  "num_item_hints": 15,
  "num_location_hints": 5,
  "num_path_hints": 0,
  "num_race_mode_dungeons": 4,
  "num_starting_triforce_shards": 3,
  "prioritize_remote_hints": false,
  "progression_battlesquid": false,
  "progression_big_octos_gunboats": false,
  "progression_combat_secret_caves": false,
  "progression_dungeons": true,
  "progression_expensive_purchases": true,
  "progression_eye_reef_chests": false,
  "progression_free_gifts": true,
  "progression_great_fairies": true,
  "progression_island_puzzles": false,
  "progression_long_sidequests": false,
  "progression_mail": false,
  "progression_minigames": false,
  "progression_misc": true,
  "progression_platforms_rafts": false,
  "progression_puzzle_secret_caves": true,
  "progression_savage_labyrinth": false,
  "progression_short_sidequests": false,
  "progression_spoils_trading": false,
  "progression_submarines": false,
  "progression_tingle_chests": false,
  "progression_treasure_charts": false,
  "progression_triforce_charts": false,
  "race_mode": true,
  "randomize_charts": false,
  "randomize_enemies": false,
  "randomize_enemy_palettes": false,
  "randomize_entrances": "Secret Caves",
  "randomize_music": false,
  "randomize_starting_island": true,
  "remove_music": false,
  "reveal_full_sea_chart": true,
  "seed_name": "yeet",
  "skip_rematch_bosses": true,
  "starting_gear": {
    "Bait Bag": 0,
    "Ballad of Gales": 0,
    "Beedle's Chart": 0,
    "Bombs": 0,
    "Boomerang": 1,
    "Cabana Deed": 1,
    "Command Melody": 0,
    "DRC Compass": 0,
    "DRC Dungeon Map": 0,
    "Deku Leaf": 0,
    "Delivery Bag": 0,
    "Din's Pearl": 0,
    "ET Compass": 0,
    "ET Dungeon Map": 0,
    "Earth God's Lyric": 0,
    "Empty Bottle": 0,
    "FF Compass": 0,
    "FF Dungeon Map": 0,
    "FW Compass": 0,
    "FW Dungeon Map": 0,
    "Farore's Pearl": 0,
    "Fill-Up Coupon": 0,
    "Ghost Ship Chart": 0,
    "Grappling Hook": 0,
    "Great Fairy Chart": 0,
    "Hero's Charm": 0,
    "Hookshot": 0,
    "Hurricane Spin": 0,
    "Iron Boots": 0,
    "Light Ring Chart": 0,
    "Maggie's Letter": 0,
    "Magic Armor": 0,
    "Moblin's Letter": 0,
    "Nayru's Pearl": 0,
    "Note to Mom": 0,
    "Octo Chart": 0,
    "Platform Chart": 0,
    "Power Bracelets": 0,
    "Progressive Bomb Bag": 0,
    "Progressive Bow": 2,
    "Progressive Magic Meter": 0,
    "Progressive Picto Box": 0,
    "Progressive Quiver": 0,
    "Progressive Shield": 0,
    "Progressive Sword": 0,
    "Progressive Wallet": 0,
    "Secret Cave Chart": 0,
    "Skull Hammer": 0,
    "Song of Passing": 0,
    "Spoils Bag": 0,
    "Submarine Chart": 0,
    "Telescope": 0,
    "Tingle Tuner": 0,
    "Tingle's Chart": 0,
    "TotG Compass": 0,
    "TotG Dungeon Map": 0,
    "WT Compass": 0,
    "WT Dungeon Map": 0,
    "Wind God's Aria": 0,
  },
  "starting_hcs": 3,
  "starting_pohs": 9,
  "swift_sail": true,
  "sword_mode": "No Starting Sword",
  "version": "1.10.0",
}
`;
